File: src/core/record_intent.py
from datetime import datetime, timezone

from pydantic import BaseModel, Field

from src.core.dns_record import Record


class RecordIntent(BaseModel):
    container_id: str = "<unknown>"
    container_name: str = "<unknown>"
    created: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    hostname: str
    force: bool = False
    record: Record

    def __eq__(self, other):
        if not isinstance(other, RecordIntent):
            return False
        
        return (
            self.container_id == other.container_id and
            self.container_name == other.container_name and
            self.hostname == other.hostname and
            self.force == other.force and
            self.record == other.record
        )
    
    def __hash__(self) -> int:
        return hash((
            self.container_id,
            self.container_name,
            self.hostname,
            self.force,
            self.record
        ))

File: internal/domain/record_intent.go
package domain

import (
	"fmt"
	"time"
)

type RecordIntent struct {
	ContainerId   string
	ContainerName string
	Created       time.Time
	Hostname      string
	Force         bool
	// TTL is the DNS record TTL in seconds. Zero means "unset" — the field is
	// omitted from the etcd value and CoreDNS applies its own default.
	TTL    uint32
	Record Record
}

func (ri RecordIntent) Render() string {
	return fmt.Sprintf("%s (container_id=%s, container_name=%s, hostname=%s, created=%s, force=%t, ttl=%d)", ri.Record.Render(), ri.ContainerId, ri.ContainerName, ri.Hostname, ri.Created.Format("2006-01-02 15:04:05"), ri.Force, ri.TTL)
}

func (ri RecordIntent) Equal(other RecordIntent) bool {
	return ri.ContainerId == other.ContainerId &&
		ri.ContainerName == other.ContainerName &&
		ri.Hostname == other.Hostname &&
		ri.Force == other.Force &&
		ri.TTL == other.TTL &&
		ri.Record.Equal(other.Record)
}

func (ri RecordIntent) Key() string {
	return fmt.Sprintf("%s|%s|%s|%t|%d|%s", ri.ContainerId, ri.ContainerName, ri.Hostname, ri.Force, ri.TTL, ri.Record.Key())
}
